### ISSUE TITLE
config library doesn't respect precedence if empty configuration is provided + command line flags

### ISSUE THREAD

**Issue by JAORMX** [stacklok/minder#2067]
### Describe the issue

In the case where the configuration looks as follows:

```yaml
---
grpc_server:
identity:
```

And some command line arguments are provided, e.g. `--grpc-host=192.168.1.7 --grpc-port=1234 --identity-url=http://localhost:1654`

The command line client does not respect the precedence order. That is, it'll randomly pick either the default for the config or the command line flag.

This is not an issue if there is no configuration file or if the configuration file actually contains the relevant values.

### What version are you using?

`main`

**Comment by Vyom-Yadav** [comment]
So, it was indeed a flaky test. I cleared the test cache and then got a few consecutive greens while testing these. @JAORMX  I would like to figure this out if you are not working on it.

**Comment by Vyom-Yadav** [comment]
@JAORMX This issue is not just with the client CLI but with the server too, so let's retitle the issue. We do not process the config files correctly. Try this test for server config:

```go
func TestReadConfigWithEmptyInputConfigAndFlags(t *testing.T) {
	t.Parallel()

	cfgstr := `---
http_server:
grpc_server:
metric_server:
`

	cfgbuf := bytes.NewBufferString(cfgstr)

	v := viper.New()
	flags := pflag.NewFlagSet("test", pflag.ContinueOnError)

	require.NoError(t, serverconfig.RegisterServerFlags(v, flags), "Unexpected error")

	require.NoError(t, flags.Parse([]string{"--http-host=foo", "--http-port=1234", "--grpc-host=bar", "--grpc-port=5678", "--metric-host=var", "--metric-port=6679"}))

	v.SetConfigType("yaml")
	require.NoError(t, v.ReadConfig(cfgbuf), "Unexpected error")

	cfg, err := serverconfig.ReadConfigFromViper(v)
	require.NoError(t, err, "Unexpected error")

	require.Equal(t, "foo", cfg.HTTPServer.Host)
	require.Equal(t, 1234, cfg.HTTPServer.Port)
	require.Equal(t, "bar", cfg.GRPCServer.Host)
	require.Equal(t, 5678, cfg.GRPCServer.Port)
	require.Equal(t, "var", cfg.MetricServer.Host)
	require.Equal(t, 6679, cfg.MetricServer.Port)
}
```

Digging into `viper` internals, I found the answer to this.

TLDR: Having keys with a `nil` value in the configuration and the viper store can give unexpected results due to how viper searches for a value for that key.

### Reason

The following config:

```
http_server:
grpc_server:
metric_server:
```
- Will add three keys: `http_server`, `grpc_server`... to the viper key store with value `nil`. 
- Now while initializing the viper store, we actually set default values for all these three flags. Well, technically not these flags, but a nested field in these structs.
- Since the direct value for these keys is `nil`, viper will try to find a value.

So viper as a last resort will read the value for `http_server` as:
```yaml
http_server:
  host: <default>
  port: <default>
```

because **we have defined** default values for it. See https://github.com/spf13/viper/blob/e36638d8786b0b58231039fc6d7db32b904dd1ba/viper.go#L1295 for value resolution. 

#### Why flaking and not failing

viper keeps a map of keys, these keys are processed in non-deterministic order. If `http_server.host` was processed before `http_server`, its value would be overridden by default viper gets for `http_server` as viper collects values of all keys before returning it to a user if we use methods like `v.getSettings(..)`.

To fix this, we should not allow keys with `nil` values which are lists or dictionaries in the config file i.e. not allow `http_server` to enter the viper key store but allow `http_server.port`, `http_server.host` in the key store.

**Comment by JAORMX** [comment]
> To fix this, we should not allow keys with nil values which are lists or dictionaries in the config file i.e. not allow http_server to enter the viper key store but allow http_server.port, http_server.host in the key store.

I think this is a good idea

I re-titled the issue. Thanks a lot for looking into this!
